File: apps/website/scripts/start.py
#!/usr/bin/env python3
"""
Launch local Supabase, Next.js development server, and Stripe webhook listener.
Press Ctrl+C to gracefully terminate all processes.
"""

import subprocess
import signal
import sys
import os
import time

pids = []
shutting_down = False


def ensure_supabase_running(project_root: str) -> None:
    """Start local Supabase if not already running."""
    print("Checking local Supabase status...")
    result = subprocess.run(
        ["yarn", "supabase", "status"],
        cwd=project_root,
        capture_output=True,
        text=True,
    )

    # Check if Supabase is running by looking for the status output
    if result.returncode != 0 or "not running" in result.stdout.lower():
        print("Starting local Supabase...")
        start_result = subprocess.run(
            ["yarn", "supabase", "start"],
            cwd=project_root,
        )
        if start_result.returncode != 0:
            print("ERROR: Failed to start Supabase. Is Docker running?")
            sys.exit(1)
        print("Supabase started successfully")
    else:
        print("Supabase already running")


def kill_tree(pid):
    """Kill a process and all its children using pkill."""
    try:
        # First try to kill child processes
        subprocess.run(["pkill", "-P", str(pid)], capture_output=True)
        # Then kill the process itself
        os.kill(pid, signal.SIGTERM)
    except (ProcessLookupError, PermissionError, OSError):
        pass


def force_kill_tree(pid):
    """Force kill a process and all its children."""
    try:
        subprocess.run(["pkill", "-9", "-P", str(pid)], capture_output=True)
        os.kill(pid, signal.SIGKILL)
    except (ProcessLookupError, PermissionError, OSError):
        pass


def cleanup(signum=None, frame=None):
    """Terminate all child processes."""
    global shutting_down
    if shutting_down:
        return
    shutting_down = True

    print("\nShutting down...")

    # Send SIGTERM to all process trees
    for pid in pids:
        kill_tree(pid)

    # Wait a moment for graceful shutdown
    time.sleep(1)

    # Force kill anything still running
    for pid in pids:
        force_kill_tree(pid)

    print("Done.")
    sys.exit(0)


def main():
    signal.signal(signal.SIGINT, cleanup)
    signal.signal(signal.SIGTERM, cleanup)

    project_root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

    # Kill any process already using port 3000
    pids_on_3000 = subprocess.run(["lsof", "-ti", ":3000"], capture_output=True, text=True).stdout.strip()
    if pids_on_3000:
        print("Killing process on port 3000...")
        subprocess.run(["kill", "-9"] + pids_on_3000.split("\n"), capture_output=True)

    # Ensure local Supabase is running before starting the app
    ensure_supabase_running(project_root)

    print("Starting Next.js development server...")
    app_proc = subprocess.Popen(
        ["yarn", "dev"],
        cwd=project_root,
        preexec_fn=os.setsid,
    )
    pids.append(app_proc.pid)

    print("Starting Stripe webhook listener...")
    webhook_proc = subprocess.Popen(
        ["yarn", "webhook"],
        cwd=project_root,
        preexec_fn=os.setsid,
    )
    pids.append(webhook_proc.pid)

    print("\nBoth processes running. Press Ctrl+C to stop.\n")

    # Wait for either process to exit
    while True:
        if app_proc.poll() is not None or webhook_proc.poll() is not None:
            break
        time.sleep(0.5)


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        cleanup()

File: apps/website/scripts/start.ts
/**
 * Launch local Supabase, Next.js development server, and Stripe webhook
 * listener. Press Ctrl+C to gracefully terminate all processes.
 */

import { spawn, spawnSync } from "node:child_process";
import { existsSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const PROJECT_ROOT = path.dirname(path.dirname(fileURLToPath(import.meta.url)));
const REPO_ROOT = path.dirname(path.dirname(PROJECT_ROOT));

const pids: number[] = [];
let shuttingDown = false;

/**
 * Start local Supabase if not already running.
 *
 * Runs against the repo-root Supabase project, not website's own
 * (apps/website/supabase) — for now this repo runs a single local
 * database, so website's local dev points at the root project. Uses
 * `yarn supabase ...` with cwd explicitly set to the repo root, rather than
 * apps/website's own `db:start`/`db:status` scripts — those run `supabase`
 * with cwd = apps/website and would target the wrong (website-local)
 * Supabase project. `supabase` is a real devDependency now (both at the
 * repo root and in apps/website), so `yarn supabase ...` resolves the local
 * CLI binary directly — no more `npx supabase` workaround.
 */
function ensureSupabaseRunning(): void {
  console.log("Checking local Supabase status...");
  const status = spawnSync("yarn", ["supabase", "status"], {
    cwd: REPO_ROOT,
    encoding: "utf-8",
  });

  const notRunning =
    status.status !== 0 || (status.stdout ?? "").toLowerCase().includes("not running");
  if (notRunning) {
    console.log("Starting local Supabase...");
    const start = spawnSync("yarn", ["supabase", "start"], {
      cwd: REPO_ROOT,
      stdio: "inherit",
    });
    if (start.status !== 0) {
      console.error("ERROR: Failed to start Supabase. Is Docker running?");
      process.exit(1);
    }
    console.log("Supabase started successfully");
  } else {
    console.log("Supabase already running");
  }
}

/** Kill a process and its direct children (SIGTERM, or SIGKILL to force). */
function killTree(pid: number, force: boolean): void {
  try {
    spawnSync("pkill", [force ? "-9" : "-TERM", "-P", String(pid)], { stdio: "ignore" });
    process.kill(pid, force ? "SIGKILL" : "SIGTERM");
  } catch {
    // process already gone
  }
}

function cleanup(): void {
  if (shuttingDown) return;
  shuttingDown = true;

  console.log("\nShutting down...");

  // Send SIGTERM to all process trees
  for (const pid of pids) killTree(pid, false);

  // Wait a moment for graceful shutdown, then force kill anything still
  // running.
  setTimeout(() => {
    for (const pid of pids) killTree(pid, true);
    console.log("Done.");
    process.exit(0);
  }, 1000);
}

function main(): void {
  process.on("SIGINT", cleanup);
  process.on("SIGTERM", cleanup);

  if (!existsSync(path.join(REPO_ROOT, "supabase", "config.toml"))) {
    console.error(
      `ERROR: expected repo root at ${REPO_ROOT}, but supabase/config.toml is missing there.`,
    );
    process.exit(1);
  }

  // Kill any process already using port 3000
  const lsof = spawnSync("lsof", ["-ti", ":3000"], { encoding: "utf-8" });
  const pidsOn3000 = (lsof.stdout ?? "").trim();
  if (pidsOn3000) {
    console.log("Killing process on port 3000...");
    spawnSync("kill", ["-9", ...pidsOn3000.split("\n")], { stdio: "ignore" });
  }

  // Ensure local Supabase is running before starting the app
  ensureSupabaseRunning();

  console.log("Starting Next.js development server...");
  const appProc = spawn("yarn", ["dev:next"], {
    cwd: PROJECT_ROOT,
    detached: true,
    stdio: "inherit",
  });
  if (appProc.pid) pids.push(appProc.pid);

  console.log("Starting Stripe webhook listener...");
  const webhookProc = spawn("yarn", ["webhook"], {
    cwd: PROJECT_ROOT,
    detached: true,
    stdio: "inherit",
  });
  if (webhookProc.pid) pids.push(webhookProc.pid);

  console.log("\nBoth processes running. Press Ctrl+C to stop.\n");

  // If either process exits on its own, tear down the other one too.
  const handleExit = () => {
    if (shuttingDown) return;
    cleanup();
  };
  appProc.on("exit", handleExit);
  webhookProc.on("exit", handleExit);
}

main();
